File: packages/feature-flags/sdk-python/feature_flags/hash.py
"""Consistent hashing utilities for percentage rollouts and variant assignment"""

import mmh3


def get_bucket(identifier: str, flag_key: str) -> int:
    """
    Get a bucket value (0-99) for a given identifier and flag key.
    Uses MurmurHash3 for fast, well-distributed hashing.
    
    The same identifier + flag_key will always get the same bucket,
    ensuring consistent user experience.
    """
    hash_input = f"{flag_key}:{identifier}"
    hash_value = mmh3.hash(hash_input, signed=False)
    return hash_value % 100


def is_in_percentage_rollout(
    identifier: str,
    flag_key: str,
    percentage: float
) -> bool:
    """
    Check if an identifier is within a percentage rollout.
    
    Args:
        identifier: User ID, session ID, or other stable identifier
        flag_key: The flag key (ensures different flags have different distributions)
        percentage: Target percentage (0-100)
    
    Returns:
        True if the identifier is within the rollout percentage
    """
    if percentage <= 0:
        return False
    if percentage >= 100:
        return True
    
    bucket = get_bucket(identifier, flag_key)
    return bucket < percentage


def assign_variant(
    identifier: str,
    flag_key: str,
    variants: list[dict]
) -> str:
    """
    Assign a variant based on weights.
    
    Args:
        identifier: User ID or session ID
        flag_key: The flag key
        variants: List of variants with 'key' and 'weight' fields
    
    Returns:
        The assigned variant key
    """
    if not variants:
        raise ValueError("No variants provided")
    
    if len(variants) == 1:
        return variants[0]["key"]
    
    bucket = get_bucket(identifier, flag_key)
    
    cumulative_weight = 0.0
    for variant in variants:
        cumulative_weight += variant["weight"]
        if bucket < cumulative_weight:
            return variant["key"]
    
    # Fallback to last variant
    return variants[-1]["key"]

File: packages/feature-flags/core/src/hash.ts
import murmur from 'murmurhash';

/**
 * Consistent hashing for percentage rollouts and variant assignment.
 * Uses MurmurHash3 for fast, well-distributed hashing.
 * 
 * The same userId + flagKey will always get the same bucket,
 * ensuring consistent user experience.
 */

/**
 * Get a bucket value (0-99) for a given identifier and flag key.
 * Used for percentage rollouts.
 */
export function getBucket(identifier: string, flagKey: string): number {
  const hashInput = `${flagKey}:${identifier}`;
  const hash = murmur.v3(hashInput);
  return Math.abs(hash) % 100;
}

/**
 * Check if an identifier is within a percentage rollout.
 * 
 * @param identifier - User ID or session ID
 * @param flagKey - The flag key (ensures different flags have different distributions)
 * @param percentage - Target percentage (0-100)
 * @returns true if the identifier is within the rollout percentage
 */
export function isInPercentageRollout(
  identifier: string,
  flagKey: string,
  percentage: number
): boolean {
  if (percentage <= 0) return false;
  if (percentage >= 100) return true;
  
  const bucket = getBucket(identifier, flagKey);
  return bucket < percentage;
}

/**
 * Assign a variant based on weights.
 * 
 * @param identifier - User ID or session ID
 * @param flagKey - The flag key
 * @param variants - Array of variants with weights (weights should sum to 100)
 * @returns The assigned variant key
 */
export function assignVariant(
  identifier: string,
  flagKey: string,
  variants: Array<{ key: string; weight: number }>
): string {
  if (variants.length === 0) {
    throw new Error('No variants provided');
  }
  
  if (variants.length === 1) {
    const first = variants[0];
    if (first) {
      return first.key;
    }
  }
  
  const bucket = getBucket(identifier, flagKey);
  
  let cumulativeWeight = 0;
  for (const variant of variants) {
    cumulativeWeight += variant.weight;
    if (bucket < cumulativeWeight) {
      return variant.key;
    }
  }
  
  // Fallback to last variant (handles rounding errors)
  const last = variants[variants.length - 1];
  return last?.key ?? variants[0]?.key ?? '';
}

/**
 * Generate a deterministic session ID from available context.
 * Used when no explicit identifier is provided.
 */
export function generateSessionId(context: {
  appName?: string;
  timestamp?: number;
  random?: string;
}): string {
  const parts = [
    context.appName ?? 'unknown',
    context.timestamp ?? Date.now(),
    context.random ?? Math.random().toString(36).substring(2, 15)
  ];
  
  return murmur.v3(parts.join(':')).toString(36);
}
